zanata-war/src/main/java/org/zanata/service/impl/GlossaryFileServiceImpl.java

```diff
@@ -127,48 +132,69 @@ private void executeCommit() {
-            // comments
-            HGlossaryTerm hGlossaryTerm =
-                    getOrCreateGlossaryTerm(to, termHLocale, glossaryTerm);
-
-            hGlossaryTerm.getComments().clear();
-
-            for (String comment : glossaryTerm.getComments()) {
-                hGlossaryTerm.getComments().add(new HTermComment(comment));
-            }
+        String resId = from.getResId();
 
-            to.getGlossaryTerms().put(termHLocale, hGlossaryTerm);
+        if(StringUtils.isBlank(resId)) {
+            resId = getResId(srcLocale, srcTerm.getContent(), from.getPos(),
+                        from.getDescription());
         }
-        glossaryDAO.makePersistent(to);
-    }
 
-    public HGlossaryEntry getOrCreateGlossaryEntry(LocaleId srcLocale,
-            String srcContent) {
-        HGlossaryEntry hGlossaryEntry =
-                glossaryDAO
-                        .getEntryBySrcLocaleAndContent(srcLocale, srcContent);
+        HGlossaryEntry hGlossaryEntry = glossaryDAO.getEntryByResId(resId);
 
         if (hGlossaryEntry == null) {
             hGlossaryEntry = new HGlossaryEntry();
             HLocale srcHLocale = localeServiceImpl.getByLocaleId(srcLocale);
             hGlossaryEntry.setSrcLocale(srcHLocale);
+            hGlossaryEntry.setResId(resId);
+            hGlossaryEntry.setSourceRef(from.getSourceReference());
+        } else {
+            // check if field in glossary entry has been updated and need for
+            // new resId
+            String newResId =
+                    getResId(srcLocale, srcTerm.getContent(), from.getPos(),
+                            from.getDescription());
+            if(resId != newResId) {
+                hGlossaryEntry.setResId(newResId);
+            }
         }
         return hGlossaryEntry;
     }
 
+    private HGlossaryEntry transferGlossaryEntryAndSave(GlossaryEntry from) {
+        HGlossaryEntry to = getOrCreateGlossaryEntry(from);
+
+        to.setSourceRef(from.getSourceReference());
+        to.setPos(from.getPos());
+        to.setDescription(from.getDescription());
+
+        for (GlossaryTerm glossaryTerm : from.getGlossaryTerms()) {
+            if(glossaryTerm.getLocale() == null || StringUtils.isBlank(
+                glossaryTerm.getContent())) {
+                continue;
+            }
+            HLocale termHLocale = localeServiceImpl.getByLocaleId(glossaryTerm
+                .getLocale());
+
+            if(termHLocale != null) {
+                // check if there's existing term
+                HGlossaryTerm hGlossaryTerm =
+                    getOrCreateGlossaryTerm(to, termHLocale, glossaryTerm);
+
+                hGlossaryTerm.setComment(glossaryTerm.getComment());
+                to.getGlossaryTerms().put(termHLocale, hGlossaryTerm);
+            } else {
+                log.warn("Language {} is not enabled in Zanata. Term will be ignored.", glossaryTerm.getLocale());
```

Comment:
Potentially this may cause thousands of log messages if user upload a file with target locale no enabled.
- [ ] should output warning outside of the loop